apps/pkeyutl.c

```diff
@@ -420,13 +421,16 @@ int pkeyutl_main(int argc, char **argv)
     if (out == NULL)
         goto end;
 
-    if (pkey_op == EVP_PKEY_OP_ENCAPSULATE) {
-        if (secoutfile == NULL) {
-            BIO_printf(bio_err, "Encapsulation requires '-secret' argument\n");
+    if (pkey_op == EVP_PKEY_OP_ENCAPSULATE
+        || pkey_op == EVP_PKEY_OP_DECAPSULATE) {
+        if (secoutfile == NULL && pkey_op == EVP_PKEY_OP_ENCAPSULATE) {
+            BIO_printf(bio_err, "KEM-based shared-secret derivation requires "
+                                "the '-secret <file>' option\n");
             goto end;
         }
-        secout = bio_open_default(secoutfile, 'w', FORMAT_BINARY);
-        if (secout == NULL)
+        /* For backwards compatibility, default decap secrets to the output */
+        if (secoutfile != NULL
+            && (secout = bio_open_owner(secoutfile, 'w', FORMAT_BINARY)) == NULL)
```

Comment:
Could you please elaborate why do you prefer this way to deal with stuff? 

I wanted to keep 2 separate values: the secret one for future use and the encapsulated one to be used by peer.  For decapsulation we have a normal input and normal output.

Reply:
If the secret is written to the `-secret` file with `-encap` it should likewise be written to the `-secret` file with `-decap`, especially because we can make sure to give the file correct ownership, and this actually lines up with the decap API, which produces a secret and no output on decap.

The backwards-compatible hack sadly means that naïve execution of `-decap` will write binary noise to terminals, messing up display (changing to graphics mode, ...). Requiring typing `stty sane^V^J` to restore a usable state.

Reply:
I'd frankly rather refuse to write to `-out`, but that is an API (CLI) break, that is perhaps not an option in a minor release.

Reply:
> If the secret is written to the `-secret` file with `-encap` it should likewise be written to the `-secret` file with `-decap`, especially because we can make sure to give the file correct ownership, and this actually lines up with the decap API, which produces a secret and no output on decap.

I'm not sure I agree. My CLI was inspired by the `-encrypt`/`-decrypt` pair with the exception we have an extra output argument in case of encapsulation. We have never been concerned about binary noise, and I also can't remember any ownership efforts.

Reply:
We make sure to correctly set the ownership of private keys containing secret material in `genpkey`, for example.
Decapsulation is not decryption, the output is a binary shared-secret for further use in KDFs, it is not useful command output.
I stroly think that alignment with `-encap` is the right thing here.

Reply:
There's at least [one user](https://groups.google.com/a/openssl.org/g/openssl-users/c/MvuKU-l51v4/m/i9pgjPcUBQAJ) who likes the proposed interface